File: 3016-minimum-number-of-pushes-to-type-word-ii/3016-minimum-number-of-pushes-to-type-word-ii.cpp
class Solution {
public:
    int minimumPushes(string word) {
        vector<int> freq(26,0);
        for(int i=0;i<word.size();i++){
            freq[word[i]-'a']++;
        }
        sort(freq.begin(),freq.end(),greater<int>());
        int curr=1,count=0,ans=0;
        for(int i=0;i<freq.size()&&freq[i]>0;i++){
            if(count!=8){
                ans+=(freq[i]*curr);
                count++;
            }
            else{
                count=0;
                curr++;
                ans+=(freq[i]*curr);
                count++;
            }
        }
        return ans;
    }
};

File: 3016-minimum-number-of-pushes-to-type-word-ii/3016-minimum-number-of-pushes-to-type-word-ii.py
class Solution:
    def minimumPushes(self, word: str) -> int:
        ans = 0
        count = Counter(word)
        maxFreq = max(count.values())
        freq = defaultdict(list)
        seen = {}
        
        for key,val in count.items():
            freq[val].append(key)
            
        k = 8
            
        while maxFreq>0:
            for val in freq[maxFreq]:
                if val not in seen:
                    seen[val]=k//8
                ans+=seen[val]*maxFreq
                k+=1
                
            maxFreq-=1
            
            
        return ans